tests/functional/test_discovery.py

```diff
@@ -0,0 +1,277 @@
+            client.test_discovery_required(Foo="bar")
+            self.assertEqual(len(stubber.requests), 1)
+            self.assertEqual(stubber.requests[0].url, "https://test.domain/")
+            self.assertEqual(
+                stubber.requests[0].headers.get('X-Amz-Target'),
+                b"awsfooservice.TestDiscoveryRequired"
+            )
+
+    def test_endpoint_discovery_optional_with_custom_endpoint(self):
+        client, http_stubber = self.create_client(
+            service_name="awsfooservice", endpoint_url="https://test.domain"
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(len(stubber.requests), 1)
+            self.assertEqual(stubber.requests[0].url, "https://test.domain/")
+            self.assertEqual(
+                stubber.requests[0].headers.get('X-Amz-Target'),
+                b"awsfooservice.TestDiscoveryOptional"
+            )
+
+    def test_endpoint_discovery_optional_disabled_with_custom_endpoint(self):
+        config = Config(endpoint_discovery_enabled=False)
+        client, http_stubber = self.create_client(
+            service_name="awsfooservice",
+            config=config,
+            endpoint_url="https://test.domain"
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(len(stubber.requests), 1)
+            self.assertEqual(stubber.requests[0].url, "https://test.domain/")
+            self.assertEqual(
+                stubber.requests[0].headers.get('X-Amz-Target'),
+                b"awsfooservice.TestDiscoveryOptional"
+            )
+
+    def test_endpoint_discovery_default_optional_endpoint(self):
+        client, http_stubber = self.create_client(service_name="awsfooservice")
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(len(stubber.requests), 1)
+
+    def test_endpoint_discovery_enabled_optional_endpoint(self):
+        config = Config(endpoint_discovery_enabled=True)
+        client, http_stubber = self.create_client(
+            service_name="awsfooservice", config=config
+        )
+        response = {
+            "Endpoints":[{
+                'Address': "https://discovered.domain",
+                'CachePeriodInMinutes': 1
+            }]
+        }
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=json.dumps(response).encode())
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(len(stubber.requests), 2)
+            self.assertEqual(
+                stubber.requests[1].url, "https://discovered.domain"
+            )
```

Comment:
It might make sense to make helper assertions to make it a bit easier to understand the assertions (e.g `assert_endpoint_used`)